sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -330,5 +350,49 @@ bool IsBooleanLiteral(SyntaxNode syntaxNode)
                 return syntaxNode.IsAnyKind(BooleanLiterals);
             }
         }
+
+        private static void CollectCoalesce(InstructionProcessedEventArgs args, HashSet<SyntaxNode> isNull, HashSet<SyntaxNode> isNotNull, HashSet<SyntaxNode> isUnknown, SemanticModel semanticModel)
+        {
+            void ProcessOperand(ExpressionSyntax operand, bool useNotNull)
+            {
+                if (operand.RemoveParentheses().IsKind(SyntaxKind.NullLiteralExpression))
+                {
+                    isNull.Add(operand);
+                    return;
+                }
+                var symbol = semanticModel.GetSymbolInfo(operand.RemoveParentheses()).Symbol;
+                if (symbol != null)
+                {
+                    if (symbol.HasConstraint(ObjectConstraint.Null, args.ProgramState))
+                    {
+                        isNull.Add(operand);
+                    }
+                    else if (useNotNull && symbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState))
+                    {
+                        isNotNull.Add(operand);
+                    }
+                    else
+                    {
+                        isUnknown.Add(operand);
+                    }
+                }
+            }
+            var rightOperands = args.Instruction.DescendantNodesAndSelf()
+                .Select(x => (x is BinaryExpressionSyntax binary && binary.IsKind(SyntaxKind.CoalesceExpression) ? binary.Right : null)
+                          ?? (x is AssignmentExpressionSyntax assign && assign.IsKind(SyntaxKindEx.CoalesceAssignmentExpression) ? assign.Right : null))
+                .WhereNotNull().ToArray();
+            foreach (var right in rightOperands)
+            {
+                //It's too late for left operands of ??= or x=x??, symbolic value was already set
+                ProcessOperand(right, false);
+            }
```

Comment:
Why do we need to look at the descendants here ? Can't we just check the current instruction ? since we will evaluate all instructions, it seems many syntax nodes will be reevaluated many times currently

Reply:
That doesn't work for `ret = a ?? null`, because instruction itself is AssignmentExpression here and there's no separate instruction for "a ?? null" in the CFG. I can check AssignmentExpression but I'm afraid that `??` can be subexpression of any kind of expression here.
DescendantNodesAndSelf will produce a lot of reevaluations and I'm hoping for small chunks here.